parlai/agents/hugging_face/llama.py

```diff
@@ -0,0 +1,380 @@
+
+
+def setup_llama_args(parser):
+    if parser is None:
+        return parser
+    parser.add_argument(
+        "--llama-size",
+        type=str,
+        default="1.5B",
+        choices=["1.5B", "7B", "30B", "65B"],
+        help="Which size model to initialize.",
+    )
+    parser.add_argument(
+        "--llama-model-dir",
+        type=str,
+        default=None,
+        help="dir to llama model and tokenizer",
+    )
+    parser.set_defaults(
+        text_truncate=1792,
+        label_truncate=256,
+        dict_maxexs=0,  # skip building dictionary
+    )
+    return parser
+
+
+############################################
+## Modules
+############################################
+
+
+class ParlaiLlamaDecoder(torch.nn.Module):
+    """
+    Llama Decoder.
+
+    This decoder is initialized with the pretrained model from Hugging Face.
+    """
+
+    def __init__(
+        self,
+        opt: Opt,
+        decoder: LlamaModel,
+        padding_idx: Optional[int] = None,
+        start_idx: Optional[int] = None,
+    ):
+        super().__init__()
+        self.model = decoder
+        self.padding_idx = padding_idx
+        self.start_idx = start_idx
+
+    def forward(
+        self, input, encoder_state, incr_state=None, sequence_classification=False
+    ):
+        attention_mask = None
+        position_ids = None
+        input_cut_start_token = False
+        if sequence_classification:
+            assert incr_state is None
+        if incr_state is None:
+            if sequence_classification:
+                # for sequence classification, ignore the input (is None anyway)
+                model_input = encoder_state
+            elif input.size(1) == 1 and (input[:, 0] == self.start_idx).all():
+                # first step generating: ignore the start token
+                model_input = encoder_state
+                input_cut_start_token = True
+            else:
+                # forced decoding: concatenate the context with the labels
+                # cut decoder_start_token:
+                if (input[:, 0] == self.start_idx).all():
+                    input = input[:, 1:]
+                    input_cut_start_token = True
+                model_input = torch.cat([encoder_state, input], dim=-1)
+
+            attention_mask = model_input != self.padding_idx
+            position_ids = (
+                attention_mask.cumsum(dim=-1, dtype=torch.int64) - 1
+            ).clamp_(min=0)
+        else:
+            # generating with continuation
+            # cut decoder_start_token
+            if (input[:, 0] == self.start_idx).all():
+                input = input[:, 1:]
+                input_cut_start_token = True
+            # generation: get the last token input
+            model_input = input[:, -1:]
+            # get the position ids
+            position_ids = (encoder_state != self.padding_idx).sum(
+                -1, True, dtype=torch.int64
+            ) - 1
```

Comment:
to make sure that position_ids starts from 0 instead of 1? like here https://github.com/facebookresearch/ParlAI/blob/main/parlai/agents/hugging_face/gpt2.py#L109-L111